doc/en/user/source/extensions/mapml/template.rst

```diff
@@ -0,0 +1,199 @@
+Templates With FreeMarker
+-------------------------
+
+MapML templates are written in `Freemarker <http://www.freemarker.org/>`_ , a Java-based template engine. The templates below are feature type specific and will not be applied in multi-layer WMS requests.  See :ref:`tutorial_freemarkertemplate` for general information about FreeMarker implementation in GeoServer.
+
+GetMap MapML HTML Preview/Layer Preview Head Stylesheet Templating
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The viewer is returned when the format includes subtype=mapml. The viewer is an HTML document that includes a head section with a link to the stylesheet. The default viewer is a simple viewer that includes a link to the default stylesheet. 
+A template can be created to insert links to whole stylesheet or actual stylesheet elements.  
+We can do this by creating a file called ``mapml-head-viewer.ftl`` in the GeoServer data directory in the directory for the layer that we wish to append links to.  For example we could create this file under ``workspaces/topp/states_shapefile/states``.  To add stylesheet links and stylesheet elements, we enter the following text inside this new file:
+
+.. code-block:: html
+
+ <!-- Added from the template -->	
+ <link rel="stylesheet" href="mystyle.css">
+ <style>
+  body {
+   background-color: linen;
+  }
+ </style>
+ <!-- End of added from the template -->
+
+This would result in a head section that would resemble:
+
+.. code-block:: html
+
+    <head>
+      <title>World Map</title>
+      <meta charset='utf-8'>
+      <script type="module"  src="http://localhost:8080/geoserver/mapml/viewer/widget/mapml-viewer.js"></script>
+      <style>
+          html, body { height: 100%; }
+          * { margin: 0; padding: 0; }
+          mapml-viewer:defined { max-width: 100%; width: 100%; height: 100%; border: none; vertical-align: middle }
+          mapml-viewer:not(:defined) > * { display: none; } nlayer- { display: none; }
+      </style>
+      <noscript>
+      <style>
+          mapml-viewer:not(:defined) > :not(layer-) { display: initial; }
+      </style>
+      </noscript>
+      <!-- Added from the template -->
+      <link rel="stylesheet" href="mystyle.css">
+      <style>
+          body {
+              background-color: linen;
+          }
+      </style>
+      <!-- End of added from the template -->
+    </head>
+
+GetMap XML Head Stylesheet Templating
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+MapML in XML format includes a map-head element that includes map-link elements to link to other resources, including map style variants.  Additional map-link elements can be added to the map-head element by creating a mapml-head.ftl template in the GeoServer data directory in the directory for the layer we wish to append map-links to.  For example we could create the head.ftl file under ``workspaces/tiger/poly_landmarks_shapefile/poly_landmarks``: 
```

Comment:
Should be `workspaces/tiger/nyc/poly_landmarks` ?  But when I locate a `mapml-head.ftl` in that directory per line 58 below, I don't see it in the output.  Not sure if it's me.

Reply:
Actually it was me and the software: the output is created, but doesn't match the documentation on lines 79-82 below.  What appears is this:

```xml
<map-link
            href="http://localhost:8080/geoserver/topp/wms?REQUEST=GetMap&amp;FORMAT=text%2Fmapml&amp;SRS=MapML%3AWGS84&amp;FORMAT_OPTIONS=%7BMAPML-WMS-FORMAT%3Dimage%2Fpng%7D&amp;BBOX=SRSEnvelope%5B-124.73142200000001%20%3A%20-66.969849%2C%2024.955967%20%3A%2049.371735%5D&amp;VERSION=1.3.0&amp;STYLES=population&amp;SERVICE=WMS&amp;WIDTH=330&amp;HEIGHT=768&amp;LAYERS=states"
            rel="style" title="templateinsertedstyle" />
        <map-style>.bbox {display:none} .population-r1-s1{fill:#4DFF4D; fill-opacity:0.7}
            .population-r2-s1{fill:#FF4D4D; fill-opacity:0.7} .population-r3-s1{fill:#4D4DFF;
            fill-opacity:0.7} .population-r4-s1{stroke-opacity:1.0; stroke-dashoffset:0;
            stroke-width:0.2; stroke:#000000; stroke-linecap:butt}
            .polygon-r1-s1{stroke-opacity:3.0; stroke-dashoffset:4; stroke-width:2.0; fill:#AAAAAA;
            fill-opacity:3.0; stroke:#DD0000; stroke-linecap:butt}</map-style>
```

So the styles are merged, and the comments don't appear. I _think_ that the template should create a new \<map-style> but maybe this is something to do with marshaling or  unmarshaling I can never remember which is which.

Reply:
In mapml-features-head.ftl I also notice that the \<map-style> elements in the template are merged into the end of the \<map-style> that is created.